### PR TITLE
BA/TA: Add missing error context - field path

### PR DESCRIPTION
**Pull request by michaljurecko**
Part of: https://keboola.atlassian.net/browse/PSGO-87

Changes:
- BA/TA API validation errors now contain context - field path, it is implemented by the `errormsg` - custom goa extension.

----------------

Vojta nahlasil, ze z chybovej hlasky nevedel zistit, co ma v requeste zle.

Pozrel som sa na to, a zistil som, ze Goa framework negeneruje pri validacii `user-defined types` kontext, tj. kde ku chybe doslo.

Namiesto `.... from body.exports[0].name` to vygeneruje iba `... from body.name`.

Vyriesil som to custom `errrormsg` goa extension:
- Da sa to jednoducho (rovnako) vyriesit aj v samotnom `goa`, mozno poslem PR, ked bude cas.

--------------------

**Predtym:**

Test:

https://github.com/keboola/keboola-as-code/blob/491488ab1a699125efb7aa329e5684b2b6549be3/test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json#L1-L5

Vygenerovany kod:
image

-------------------------

**Potom:**

Test:

https://github.com/keboola/keboola-as-code/blob/6a986a4fa307a2346038a26f939ea4a8111a6a6f/test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json#L1-L5

Vygenerovany kod:
image

### CODE REVIEW

File: test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "statusCode": 400,
+  "error": "buffer.missingField",
+  "message": "\"name\" is missing from body; \"mapping\" is missing from body."
+}
```
Comment (michaljurecko) [inline test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json]: Pridany E2E test, tj. nespravny stav.

File: internal/pkg/service/buffer/api/gen/http/buffer/server/types.go [hunk 2]
```diff
@@ -1132,33 +1134,34 @@ func NewGetTaskPayload(taskID string, storageAPIToken string) *buffer.GetTaskPay
...
+			err = goa.MergeErrors(err, goa.InvalidLengthError(strings.Join(append(errContext, "name"), "."), *body.Name, utf8.RuneCountInString(*body.Name), 40, false))
 		}
 	}
-	for _, e := range body.Exports {
+	for i, e := range body.Exports {
+		errContext := append(errContext, fmt.Sprintf(`exports[%d]`, i))
 		if e != nil {
-			if err2 := ValidateCreateExportDataRequestBody(e); err2 != nil {
+			if err2 := ValidateCreateExportDataRequestBody(e, errContext); err2 != nil {
 				err = goa.MergeErrors(err, err2)
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/api/gen/http/buffer/server/types.go:1165]: Ako sa rekurzivne vo vygenerovanom kode validuju chyby, tak sa generuje `errContext`, ktory definuje, ako sme sa k danej hodnote dostali.

File: test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
   "statusCode": 400,
   "error": "buffer.missingField",
-  "message": "\"name\" is missing from body; \"mapping\" is missing from body."
+  "message": "\"name\" is missing from body.exports[0]; \"mapping\" is missing from body.exports[0]"
```
Comment (michaljurecko) [inline test/buffer/api/receivers/create-013-incomplete-export-400/001-create/expected-response.json]: A upraveny E2E, uz spravny vystup.

File: test/buffer/api/exports/create-008-empty-columns-400/001-create-400/expected-response.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
   "statusCode": 400,
   "error": "buffer.invalidLength",
-  "message": "Length of body.columns must be greater or equal than 1 but got value []*server.ColumnRequestBody{} (len=0)"
+  "message": "Length of body.mapping.columns must be greater or equal than 1 but got value []*server.ColumnRequestBody{} (len=0)"
```
Comment (michaljurecko) [inline test/buffer/api/exports/create-008-empty-columns-400/001-create-400/expected-response.json:4]: Fixnuty este jeden existujuci test.

**Review by jprochazk — APPROVED** [review]

**Comment by michaljurecko** [comment]
@zajca mohol by si prosim dat approve? Skontroloval to Honza este v piatok, ale uz nie je v nasom GitHub-e, tak to nemozem mergnut.

**Review by zajca — APPROVED** [review]

**Pull request merged**